std/json.zig

```diff
@@ -0,0 +1,1308 @@
+                    p.pushStack(try p.parseString(allocator, token, input, i));
+                },
+                TokenId.Number => {
+                    p.pushStack(try p.parseNumber(token, input, i));
+                },
+                TokenId.True => {
+                    p.pushStack(Value { .Bool = true });
+                },
+                TokenId.False => {
+                    p.pushStack(Value { .Bool = false });
+                },
+                TokenId.Null => {
+                    p.pushStack(Value.Null);
+                },
+                TokenId.ObjectEnd, TokenId.ArrayEnd => {
+                    unreachable;
+                },
+            },
+        }
+    }
+
+    fn pushToParent(p: &JsonParser, value: &const Value) !void {
+        switch (p.stack[p.stack_used - 1]) {
+            // Object Parent -> [ ..., object, <key>, value ]
+            Value.String => |key| {
+                p.stack_used -= 1;
+
+                var object = &p.stack[p.stack_used - 1].Object;
+                _ = try object.put(key, value);
+                p.state = State.ObjectKey;
+            },
+            // Array Parent -> [ ..., <array>, value ]
+            Value.Array => |*array| {
+                try array.append(value);
+                p.state = State.ArrayValue;
+            },
+            else => {
+                unreachable;
+            },
+        }
+    }
+
+    fn pushStack(p: &JsonParser, value: &const Value) void {
+        p.stack[p.stack_used] = *value;
+        p.stack_used += 1;
+    }
+
+    fn parseString(p: &JsonParser, allocator: &Allocator, token: &const Token, input: []const u8, i: usize) !Value {
+        // TODO: We don't strictly have to copy values which do not contain any escape
+        // characters if flagged with the option.
+        const slice = token.slice(input, i);
+        return Value { .String = try mem.dupe(p.allocator, u8, slice) };
+    }
+
+    fn parseNumber(p: &JsonParser, token: &const Token, input: []const u8, i: usize) !Value {
+        return if (token.number_is_integer)
+            Value { .Integer = try std.fmt.parseInt(i64, token.slice(input, i), 10) }
+        else
+            @panic("TODO: fmt.parseFloat not yet implemented")
+        ;
+    }
+};
+
+const debug = std.debug;
+
+test "json parser dynamic" {
+    var p = JsonParser.init(std.debug.global_allocator, false);
+
+    const s =
+      \\{
+      \\  "Image": {
+      \\      "Width":  800,
+      \\      "Height": 600,
+      \\      "Title":  "View from 15th Floor",
+      \\      "Thumbnail": {
+      \\          "Url":    "http://www.example.com/image/481989943",
+      \\          "Height": 125,
+      \\          "Width":  100
+      \\      },
+      \\      "Animated" : false,
+      \\      "IDs": [116, 943, 234, 38793]
+      \\    }
+      \\}
+      ;
+
+    var tree = try p.parse(s);
+    tree.deinit();
+    var root = tree.root;
+
+    var image = (??root.Object.get("Image")).value;
```

Comment:
We could expose some helper methods on `Value` to make accessing nested objects slightly easier. Non-essential for the moment.